receivers/mqtt/mqtt.go

```diff
@@ -0,0 +1,135 @@
+package mqtt
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	mqttLib "github.com/eclipse/paho.mqtt.golang"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/types"
+
+	"github.com/grafana/alerting/logging"
+	"github.com/grafana/alerting/receivers"
+	"github.com/grafana/alerting/templates"
+)
+
+type Client interface {
+	Connect() mqttLib.Token
+	Publish(topic string, qos byte, retained bool, payload interface{}) mqttLib.Token
+	Disconnect(quiesce uint)
+}
+
+type Notifier struct {
+	*receivers.Base
+	log      logging.Logger
+	tmpl     *templates.Template
+	settings Config
+	client   Client
+}
+
+func defaultClientFactory(opts *mqttLib.ClientOptions) Client {
+	return mqttLib.NewClient(opts)
+}
+
+func New(cfg Config, meta receivers.Metadata, template *templates.Template, logger logging.Logger, clientFactory func(opts *mqttLib.ClientOptions) Client) *Notifier {
+	if clientFactory == nil {
+		clientFactory = defaultClientFactory
+	}
+
+	opts := mqttLib.NewClientOptions().
+		AddBroker(cfg.BrokerURL).
+		SetClientID(cfg.ClientID).
+		SetUsername(cfg.Username).
+		SetPassword(cfg.Password)
+
+	if cfg.InsecureSkipVerify {
+		tlsCfg := tls.Config{
+			InsecureSkipVerify: true,
+		}
+		opts.SetTLSConfig(&tlsCfg)
+	}
+
+	return &Notifier{
+		Base:     receivers.NewBase(meta),
+		log:      logger,
+		tmpl:     template,
+		settings: cfg,
+		client:   clientFactory(opts),
+	}
+}
+
+// mqttMessage defines the JSON object send to an MQTT broker.
+type mqttMessage struct {
+	*templates.ExtendedData
+
+	// The protocol version.
+	Version  string `json:"version"`
+	GroupKey string `json:"groupKey"`
+	Message  string `json:"message"`
+}
+
+func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	n.log.Debug("Sending an MQTT message")
+
+	msg, err := n.buildMessage(ctx, as...)
+	if err != nil {
+		return false, err
+	}
+
+	if token := n.client.Connect(); token.Wait() && token.Error() != nil {
+		n.log.Error("Failed to connect to MQTT broker", "error", token.Error())
+		return false, fmt.Errorf("Failed to connect to MQTT broker: %w", token.Error())
+	}
+
+	if token := n.client.Publish(n.settings.Topic, 0, false, string(msg)); token.Wait() && token.Error() != nil {
```

Comment:
Publish [accepts](https://github.com/eclipse/paho.mqtt.golang/blob/714f7c0231294ec4144f0e0e5fc5b43a6d430d2f/client.go#L82) `QoS` and `retained` which are set to `0` and `false` here. I started adding them too, but I'll do separate PRs later if that's fine, as the current PRs (to Grafana, alerting and the terraform provider) are becoming large.